scripts/install_third_party_test.py

```diff
@@ -0,0 +1,376 @@
+            pass
+        def mock_extractall(unused_self, path):
+            self.check_function_calls['extractall_is_called'] = True
+        # pylint: enable=unused-argument
+
+        self.unzip_swap = self.swap(
+            install_third_party, 'TMP_UNZIP_PATH', MOCK_TMP_UNZIP_PATH)
+        self. dir_exists_swap = self.swap(
+            common, 'ensure_directory_exists', mock_ensure_directory_exists)
+        self.exists_swap = self.swap(os.path, 'exists', mock_exists)
+        self.remove_swap = self.swap(os, 'remove', mock_remove)
+        self.rename_swap = self.swap(os, 'rename', mock_rename)
+        self.url_retrieve_swap = self.swap(
+            python_utils, 'url_retrieve', mock_url_retrieve)
+        self.extract_swap = self.swap(
+            zipfile.ZipFile, 'extractall', mock_extractall)
+
+    def test_download_files_with_invalid_source_filenames(self):
+        with self.assertRaises(AssertionError):
+            install_third_party.download_files(
+                'source_url', 'target_dir', 'invalid source filename')
+
+    def test_download_files_with_valid_source_filenames(self):
+        check_file_downloads = {
+            'target_dir/file1': False,
+            'target_dir/file2': False
+        }
+        expected_check_file_downloads = {
+            'target_dir/file1': False,
+            'target_dir/file2': True
+        }
+
+        def mock_exists(path):
+            if path == 'target_dir/file1':
+                return True
+            return False
+        def mock_url_retrieve(unused_url, filename):
+            check_file_downloads[filename] = True
+
+        exists_swap = self.swap(os.path, 'exists', mock_exists)
+        url_retrieve_swap = self.swap(
+            python_utils, 'url_retrieve', mock_url_retrieve)
+        with self.dir_exists_swap, exists_swap, url_retrieve_swap:
+            install_third_party.download_files(
+                'source_url', 'target_dir', ['file1', 'file2'])
+        self.assertEqual(check_file_downloads, expected_check_file_downloads)
+
+    def test_download_and_unzip_files_without_exception(self):
+        exists_arr = []
+        self.check_function_calls['url_open_is_called'] = False
+        self.expected_check_function_calls['url_open_is_called'] = False
+        def mock_exists(unused_path):
+            exists_arr.append(False)
+            return False
+
+        exists_swap = self.swap(os.path, 'exists', mock_exists)
+
+        with exists_swap, self.dir_exists_swap, self.url_retrieve_swap:
+            with self.remove_swap, self.rename_swap, self.unzip_swap:
+                with self.extract_swap:
+                    install_third_party.download_and_unzip_files(
+                        'source url', 'target dir', 'zip root', 'target root')
+        self.assertEqual(
+            self.check_function_calls, self.expected_check_function_calls)
+        self.assertEqual(exists_arr, [False])
+
+    def test_download_and_unzip_files_with_exception(self):
```

Comment:
I don't see a check for the exception here...

Reply:
In case of an exception in the function `download_and_unzip_files`, the exception is caught and a [different code](https://github.com/oppia/oppia/pull/7790/files#diff-e339ec800b0bc19a36876f3ded686b2dR133) is executed for download of files. While executing that code, a few functions are called which are not called if exception does not occur. So, in this test, I just check if those functions are called.